addons/web_editor/controllers/bus.py

```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import re
+
+from odoo.http import request
+from odoo.addons.bus.controllers.main import BusController
+
+
+class EditorCollaborationController(BusController):
+    # ---------------------------
+    # Extends BUS Controller Poll
+    # ---------------------------
+    def _poll(self, dbname, channels, last, options):
+        if request.session.uid:
+            # Do not alter original list.
+            channels = list(channels)
+            for channel in channels:
+                if isinstance(channel, str):
+                    match = re.match(r'editor_collaboration:(\w+(?:.\w+)*):(\w+):([\d]+)', channel)
```

Comment:
Instead of using a pseudo-channel-uuid that is parsed just for the access check, I propose to simplify this and use a real opaque channel-id, generated based on a crypto hmac signature, derived from the database secret key.
In order to permit access checks, it would be generated server-side only after checking that the user has the necessary access. And then it could be used everywhere on the client-side like a "normal" opaque channel (like for livechat).

If we want to reinforce the security and avoid giving the user a "permanent" channel that they could continue to use for months, even if their access is blocked, we can add a timestamp in the signature, and return the timestamp along with the hash. This way the client-side knows that it may need to ask for a new channel.

some pseudo-code to explain:

```py
# the method to obtain the channel ID, located on some model,
# perhaps bus.bus, and called from a controller that may need it?
def _editor_serialization_channel(model_name, field_name, res_id):
    # verify access to the edition channel
    if not self.env.user.has_group('base.group_user'):
        raise AccessDenied()
    recs = self.env[model_name].browse(res_id)
    recs.check_access_rights('write')
    recs.check_access_rule('write')
    (...)

    # generate unique channel id as HMAC
    # looks like this: 3e1ef9e86...3036d65 (64 chars)
    # timestamp is only valid 1 week, then you have to renew it
    exp_timestamp = int(time.time() // (7 * 24 * 3600)) + 1  # +1 for next week
    sig = odoo.tools.hmac(
        self.env, 'editor_serialization', f"{model_name}:{field_name}:{res_id}:{exp_timestamp}")

    # The result contains the signature (channel_id), followed by the expiration
    # timestamp for that channel.
    # If the result of calling `int(time.time() // (7 * 24 * 3600))` on
    # the client side gives a different timestamp, it means the channel has expired, and
    # a new one should be requested.
    return f"{sig}:{exp_timestamp}"
```

Reply:
> Instead of using a pseudo-channel-uuid that is parsed just for the access check, I propose to simplify this and use a real opaque channel-id, generated based on a crypto hmac signature, derived from the database secret key.
> In order to permit access checks, it would be generated server-side only after checking that the user has the necessary access. And then it could be used everywhere on the client-side like a "normal" opaque channel (like for livechat).
> 
> If we want to reinforce the security and avoid giving the user a "permanent" channel that they could continue to use for months, even if their access is blocked, we can add a timestamp in the signature, and return the timestamp along with the hash. This way the client-side knows that it may need to ask for a new channel.
> 
> some pseudo-code to explain:
> 
> ```python
> # the method to obtain the channel ID, located on some model,
> # perhaps bus.bus, and called from a controller that may need it?
> def _editor_serialization_channel(model_name, field_name, res_id):
>     # verify access to the edition channel
>     if not self.env.user.has_group('base.group_user'):
>         raise AccessDenied()
>     recs = self.env[model_name].browse(res_id)
>     recs.check_access_rights('write')
>     recs.check_access_rule('write')
>     (...)
> 
>     # generate unique channel id as HMAC
>     # looks like this: 3e1ef9e86...3036d65 (64 chars)
>     # timestamp is only valid 1 week, then you have to renew it
>     exp_timestamp = int(time.time() // (7 * 24 * 3600)) + 1  # +1 for next week
>     sig = odoo.tools.hmac(
>         self.env, 'editor_serialization', f"{model_name}:{field_name}:{res_id}:{exp_timestamp}")
> 
>     # The result contains the signature (channel_id), followed by the expiration
>     # timestamp for that channel.
>     # If the result of calling `int(time.time() // (7 * 24 * 3600))` on
>     # the client side gives a different timestamp, it means the channel has expired, and
>     # a new one should be requested.
>     return f"{sig}:{exp_timestamp}"
> ```

As this feature is urgent, does it seems currently secure enough to apply your suggestion on a subsequent PR ?

Reply:
After a conversation with AL, this solution is interesting but has two flaws.
1. there is an edge case when one person connecting to a channel in week X will be disconnected from a person in week X+1

2. it is impossible to change the access right of a person in real time as the person could reuse a token for one week

AL gave his validation to remain as is and find an improved solution later.